feat(input[range]): support step

Narretz opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
feat

**What is the current behavior? (You can also link to an open issue here)**
Input range doesn't handle the step attribute wrt to validation and model updates

**Does this PR introduce a breaking change?**
No

**TODO**
- [ ] Add tests for step + min + max
- [x] Add docs
- [x] Find out exactly what is broken in Chrome

**Other information**:

Wrt the Chrome bug mentioned in https://github.com/angular/angular.js/pull/14870#issuecomment-235829620, it looks like setting a number isn't the problem, but that Chrome has a bug when you change the step value multiple times. Not sure yet, though.

Narretz committed:
feat(input[range]): support step

Step support works like min / max, but with the following caveat.
Currently, only Firefox fully implements the spec. Other browsers
(Chrome, Safari, Edge) have issues when the step value changes
after the input has been changed. They do not adjust the input value
to a valid value, but instead set the stepMismatch validity state.

Angular will take this validity state, and forward it as the ngModel
"step" error. Adjusting the error ourselves would add too much code,
as the logic is quite involved.

Narretz commented:
@gkalpak I can't replicate the bug you described in https://github.com/angular/angular.js/pull/14870#issuecomment-235829620. Here's my repro: http://plnkr.co/edit/cBc2YAySJzD8y0NukAEQ?p=preview
What happens instead is that Chrome sets the value to "60", which is also weird, as "40" (like Firefox sets it) is much closer to "45" (the previous value).

There's also another test case (fn angTestCase()) that shows how Chrome stops adjusting the element value based on the step when the value has been modified previously. 

I think I'll raise both bugs in the Webkit tracker, but I'm totally prepared that this is following the spec in some obscure way ...

gkalpak commented:
I can't reproduce it on Chrome v52 either. (I was on v51 when I reproduced it. It got probably fixed,  introducing another bug :wink:)

Narretz commented:
Reported as https://bugs.chromium.org/p/chromium/issues/detail?id=633075 (stops adjusting) and https://bugs.chromium.org/p/chromium/issues/detail?id=633074 (adjusts to wrong value)

petebacondarwin commented on src/ng/directive/input.js:
```diff
@@ -1570,6 +1574,64 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+    stepChange(attr.step);
+    attr.$observe('step', stepChange);
+  }
+
```
There is a lot of similar (almost duplicate) code here. In the interests of keeping the code size down is it possible to move this into a helper function? Perhaps:

```js
setupHandlers(type, attrType, noopValidationFn, validatorFn, changeFn) {
  if (attrType) {
    ctrl.$validators[type] = attributeType === type && supportsRange ? noopValidatorFn : validatorFn;
    if (attrType === type) {
      element.attr(type, attr[type]);
    }
    changeFn(attr[type]);
    attr.$observe(type, changeFn);
  }
}
```

petebacondarwin commented:
One refactoring comment. Other than that LGTM

Narretz committed:
feat(input[number]): support step

input[number] will now set the step error if the input value
(ngModel $viewValue) does not fit the step constraint set in the step / ngStep attribute.

Fixes #10597

Narretz commented:
@petebacondarwin Sorry Pete, the current state of the PR is outdated :o I just pushed the updates. 
It also included step support for input number.

The main difference for range to before is that we don't support ngStep, because the browser would still use the default step of 1, and we don't test the step change updates anymore, because not all browsers implement them.

petebacondarwin commented on src/ng/directive/input.js:
```diff
@@ -1642,6 +1690,21 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
   }
 
+  function stepChange(val) {
+    stepVal = parseNumberAttrVal(val);
```
Perhaps do this after the check for NaN below to save unnecessary computation?

petebacondarwin replied:
Oh - this is a side-effect? OK

petebacondarwin commented:
I feel sad that we don't support `ngStep` for native range inputs. It feels confusing for the developer to have it but only for non-native ranges.

Narretz commented:
That's why I didn't document it at all. Because it's not really useful.

Narretz commented:
Landed as e1da4bed8e291003d485a8ad346ab80bed8ae2e3 and 9a8b8aaa9641bfc85d245c4355fa5481285b593f
